subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/validation/MavenPublicationErrorChecker.java

```diff
@@ -43,14 +48,56 @@ public abstract class MavenPublicationErrorChecker extends PublicationErrorCheck
      * @throws PublishException if the artifacts are modified
      */
     public static void checkThatArtifactIsPublishedUnmodified(PublishArtifact source, DefaultMavenArtifactSet mainArtifacts) {
+        // Do we want to allow _additional_ artifacts to published? If not, we need to completely associate the source and artifacts
+        // at a higher level. As-is, this just verifies that none are removed from the components artifacts.
+        MavenArtifact bestMatch = null;
+        Set<Difference> bestMatchDifferences = null;
         for (MavenArtifact mavenArtifact : mainArtifacts) {
-            if (source.getFile().equals(mavenArtifact.getFile())
-                && source.getExtension().equals(mavenArtifact.getExtension())
-                && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+            EnumSet<Difference> difference = EnumSet.noneOf(Difference.class);
+            if (!source.getFile().equals(mavenArtifact.getFile())) {
+                difference.add(Difference.FILE);
+            }
+            if (!source.getExtension().equals(mavenArtifact.getExtension())) {
+                difference.add(Difference.EXTENSION);
+            }
+            if (!Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+                difference.add(Difference.CLASSIFIER);
+            }
+            // If it's all equal, we found a matching artifact that is being published
+            if (difference.isEmpty()) {
                 return;
             }
+            if (bestMatch == null || difference.size() < bestMatchDifferences.size()) {
+                bestMatch = mavenArtifact;
+                bestMatchDifferences = difference;
+            }
         }
+        if (bestMatch == null) {
+            throw new PublishException("Cannot publish module metadata because it does not contain any artifacts.");
+        }
+        String differences = getDifferences(source, bestMatch, bestMatchDifferences);
+        throw new PublishException("Cannot publish module metadata because an artifact from a component has been removed. The best match had these problems:\n" + differences);
+    }
+
+    private static String getDifferences(PublishArtifact source, MavenArtifact bestMatch, Set<Difference> bestMatchDifferences) {
+        return bestMatchDifferences.stream().map(diff -> {
+            switch (diff) {
+                case FILE:
+                    return "- file differs from component artifact: (artifact) " + source.getFile() + " != (best match) " + bestMatch.getFile();
```

Comment:
❗ This prints the absolute path of the file in the failure, I think the relative path from project root or even just the file name would be better.  Leaking absolute path info into logs should probably be avoided.

Reply:
I don't think that's worth the effort when there are plenty of other ways this could happen that are far more likely, such as a simple compilation failure.